### PR TITLE
use xtasks for release builds and man page generation

### PR DESCRIPTION
**Pull request by b5**
This PR adds support for manual pages generated from `clap` documentation, and the [cargo xtask pattern](https://github.com/matklad/cargo-xtask) for coding up low-overhead automation in Rust.

### CODE REVIEW

File: xtask/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "xtask"
+version = "0.1.0"
+edition = "2021"
```
Comment (dignifiedquire) [inline xtask/Cargo.toml:4]: this should be marked as private to avoid accidentally trying to publish this crate

File: xtask/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+fn main() {
+    if let Err(e) = try_main() {
+        eprintln!("{}", e);
+        std::process::exit(-1);
+    }
+}
+
+fn try_main() -> Result<(), DynError> {
+    let task = env::args().nth(1);
```
Comment (dignifiedquire) [inline xtask/src/main.rs:18]: why not use clap instead?

File: xtask/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+fn dist_manpage() -> Result<(), DynError> {
+    let app = iroh::run::Cli::command();
+
+    // TODO(b5): This only generates the man page for the iroh command
+    // need to recursively handle all subcommands, generating man files
+    // for them as well
+
+    let man = clap_mangen::Man::new(app);
+    let mut buffer: Vec<u8> = Default::default();
+    man.render(&mut buffer)?;
```
Comment (dignifiedquire) [inline xtask/src/main.rs:88]: instead of writing to memory you can create and open the file, create a bufwriter aroubd it and the write to it

**Comment by dignifiedquire** [comment]
please remove .ds_store :)

File: xtask/src/main.rs [hunk 4]
```diff
@@ -6,32 +6,33 @@ use std::{
     process::{Command, Stdio},
 };
 
+fn cli() -> clap::Command {
+    clap::Command::new("xtasks")
```
Comment (dignifiedquire) [inline xtask/src/main.rs:10]: please use the struct form for command definitions, it‘s considerably easier to maintain and reason about

File: xtask/src/main.rs [hunk 5]
```diff
@@ -6,32 +6,33 @@ use std::{
...
-    let task = env::args().nth(1);
-    match task.as_deref() {
-        Some("dist") => dist()?,
-        Some("man") => dist_manpage()?,
-        _ => print_help(),
+fn run_subcommand(matches: clap::ArgMatches) -> Result<()> {
+    match matches.subcommand() {
+        Some(("dist", _)) => dist()?,
+        Some(("man", _)) => dist_manpage()?,
+        _ => unreachable!(), // If all subcommands are defined above, anything else is unreachabe!()
```
Comment (dignifiedquire) [inline xtask/src/main.rs:31]: None is very possible if no subcommand is given.

**Review by ramfox — DISMISSED** [review]

File: xtask/src/main.rs (hunk 4 above)
Comment (b5) [reply xtask/src/main.rs:10]: for those of you who are well versed in rust compiler magic, sure. I personally find this form more readable.

That said, I'm sure I'll come to learn this compiler magic in due time, and will defer to you on all things Rust. Will change to the struct form

File: xtask/src/main.rs (hunk 5 above)
Comment (b5) [reply xtask/src/main.rs:31]: nah, that's caught by the combination of `cli().get_matches();` and the root command `arg_required_else_help`

Comment (dignifiedquire) [reply xtask/src/main.rs]: great :)

File: xtask/src/main.rs (hunk 4 above)
Comment (dignifiedquire) [reply xtask/src/main.rs]: I am looking forward to once your brain has embraced the strength and statics of types 💙

**Review by dignifiedquire — APPROVED** [review]
Last note, maybe add something to our development docs and/or the readme about this?

**Comment by b5** [comment]
yeah, I plan to put a little more work into the actual things that are automated (like generating _all_ release binaries) before listing this tool in `DEVELOPERS.md`. Also want to talk to @Arqu about how this may-or-may-not intersect with CI tasks

**Pull request merged**